### ISSUE TITLE
Compiler docs not building due to stability

### ISSUE THREAD

**Issue by Manishearth** [rust-lang/rust#40880]
env_logger (pulled in by cargo) links to `log`, which needs rustc_private. The doc step needs to be able to sidestep this.

<details>
```
 Documenting env_logger v0.4.2
error: use of unstable library feature 'rustc_private' (see issue #27812)
   --> /Users/manishearth/.cargo/registry/src/github.com-1ecc6299db9ec823/env_logger-0.4.2/src/lib.rs:138:1
    |
138 | extern crate log;
    | ^^^^^^^^^^^^^^^^^
    |
    = help: add #![feature(rustc_private)] to the crate attributes to enable

error: use of unstable library feature 'rustc_private' (see issue #27812)
   --> /Users/manishearth/.cargo/registry/src/github.com-1ecc6299db9ec823/env_logger-0.4.2/src/lib.rs:145:11
    |
145 | use log::{Log, LogLevel, LogLevelFilter, LogRecord, SetLoggerError, LogMetadata};
    |           ^^^
    |
    = help: add #![feature(rustc_private)] to the crate attributes to enable

error: use of unstable library feature 'rustc_private' (see issue #27812)
   --> /Users/manishearth/.cargo/registry/src/github.com-1ecc6299db9ec823/env_logger-0.4.2/src/lib.rs:145:16
    |
145 | use log::{Log, LogLevel, LogLevelFilter, LogRecord, SetLoggerError, LogMetadata};
    |                ^^^^^^^^
    |
    = help: add #![feature(rustc_private)] to the crate attributes to enable

error: use of unstable library feature 'rustc_private' (see issue #27812)
   --> /Users/manishearth/.cargo/registry/src/github.com-1ecc6299db9ec823/env_logger-0.4.2/src/lib.rs:145:26
    |
145 | use log::{Log, LogLevel, LogLevelFilter, LogRecord, SetLoggerError, LogMetadata};
    |                          ^^^^^^^^^^^^^^
    |
    = help: add #![feature(rustc_private)] to the crate attributes to enable

error: use of unstable library feature 'rustc_private' (see issue #27812)
   --> /Users/manishearth/.cargo/registry/src/github.com-1ecc6299db9ec823/env_logger-0.4.2/src/lib.rs:145:42
    |
145 | use log::{Log, LogLevel, LogLevelFilter, LogRecord, SetLoggerError, LogMetadata};
    |                                          ^^^^^^^^^
    |
    = help: add #![feature(rustc_private)] to the crate attributes to enable

error: use of unstable library feature 'rustc_private' (see issue #27812)
   --> /Users/manishearth/.cargo/registry/src/github.com-1ecc6299db9ec823/env_logger-0.4.2/src/lib.rs:145:53
    |
145 | use log::{Log, LogLevel, LogLevelFilter, LogRecord, SetLoggerError, LogMetadata};
    |                                                     ^^^^^^^^^^^^^^
    |
    = help: add #![feature(rustc_private)] to the crate attributes to enable

error: use of unstable library feature 'rustc_private' (see issue #27812)
   --> /Users/manishearth/.cargo/registry/src/github.com-1ecc6299db9ec823/env_logger-0.4.2/src/lib.rs:145:69
    |
145 | use log::{Log, LogLevel, LogLevelFilter, LogRecord, SetLoggerError, LogMetadata};
    |                                                                     ^^^^^^^^^^^
    |
    = help: add #![feature(rustc_private)] to the crate attributes to enable

error: use of unstable library feature 'rustc_private' (see issue #27812)
   --> /Users/manishearth/.cargo/registry/src/github.com-1ecc6299db9ec823/env_logger-0.4.2/src/lib.rs:159:21
    |
159 |     format: Box<Fn(&LogRecord) -> String + Sync + Send>,
    |                     ^^^^^^^^^
    |
    = help: add #![feature(rustc_private)] to the crate attributes to enable

error: use of unstable library feature 'rustc_private' (see issue #27812)
   --> /Users/manishearth/.cargo/registry/src/github.com-1ecc6299db9ec823/env_logger-0.4.2/src/lib.rs:198:21
    |
198 |     format: Box<Fn(&LogRecord) -> String + Sync + Send>,
    |                     ^^^^^^^^^
    |
    = help: add #![feature(rustc_private)] to the crate attributes to enable

error: use of unstable library feature 'rustc_private' (see issue #27812)
   --> /Users/manishearth/.cargo/registry/src/github.com-1ecc6299db9ec823/env_logger-0.4.2/src/lib.rs:325:5
    |
325 |       fn enabled(&self, metadata: &LogMetadata) -> bool {
    |  _____^ starting here...
326 | |         self.enabled(metadata.level(), metadata.target())
327 | |     }
    | |_____^ ...ending here
    |
    = help: add #![feature(rustc_private)] to the crate attributes to enable

error: use of unstable library feature 'rustc_private' (see issue #27812)
   --> /Users/manishearth/.cargo/registry/src/github.com-1ecc6299db9ec823/env_logger-0.4.2/src/lib.rs:329:5
    |
329 |     fn log(&self, record: &LogRecord) {
    |     ^
    |
    = help: add #![feature(rustc_private)] to the crate attributes to enable

error: use of unstable library feature 'rustc_private' (see issue #27812)
   --> /Users/manishearth/.cargo/registry/src/github.com-1ecc6299db9ec823/env_logger-0.4.2/src/lib.rs:324:6
    |
324 | impl Log for Logger {
    |      ^^^
    |
    = help: add #![feature(rustc_private)] to the crate attributes to enable

error: use of unstable library feature 'rustc_private' (see issue #27812)
   --> /Users/manishearth/.cargo/registry/src/github.com-1ecc6299db9ec823/env_logger-0.4.2/src/lib.rs:346:12
    |
346 |     level: LogLevelFilter,
    |            ^^^^^^^^^^^^^^
    |
    = help: add #![feature(rustc_private)] to the crate attributes to enable

error: use of unstable library feature 'rustc_private' (see issue #27812)
   --> /Users/manishearth/.cargo/registry/src/github.com-1ecc6299db9ec823/env_logger-0.4.2/src/lib.rs:354:29
    |
354 | pub fn init() -> Result<(), SetLoggerError> {
    |                             ^^^^^^^^^^^^^^
    |
    = help: add #![feature(rustc_private)] to the crate attributes to enable

error: use of unstable library feature 'rustc_private' (see issue #27812)
   --> /Users/manishearth/.cargo/registry/src/github.com-1ecc6299db9ec823/env_logger-0.4.2/src/lib.rs:386:32
    |
386 |                     Err(_) => (LogLevelFilter::max(), Some(part0)),
    |                                ^^^^^^^^^^^^^^^^^^^
    |
    = help: add #![feature(rustc_private)] to the crate attributes to enable

error: use of unstable library feature 'rustc_private' (see issue #27812)
   --> /Users/manishearth/.cargo/registry/src/github.com-1ecc6299db9ec823/env_logger-0.4.2/src/lib.rs:389:47
    |
389 |             (Some(part0), Some(""), None) => (LogLevelFilter::max(), Some(part0)),
    |                                               ^^^^^^^^^^^^^^^^^^^
    |
    = help: add #![feature(rustc_private)] to the crate attributes to enable

error: use of unstable library feature 'rustc_private' (see issue #27812)
   --> /Users/manishearth/.cargo/registry/src/github.com-1ecc6299db9ec823/env_logger-0.4.2/src/lib.rs:207:40
    |
207 |             format: Box::new(|record: &LogRecord| {
    |                                        ^^^^^^^^^
    |
    = help: add #![feature(rustc_private)] to the crate attributes to enable

error: use of unstable library feature 'rustc_private' (see issue #27812)
   --> /Users/manishearth/.cargo/registry/src/github.com-1ecc6299db9ec823/env_logger-0.4.2/src/lib.rs:220:26
    |
220 |                   level: LogLevelFilter) -> &mut Self {
    |                          ^^^^^^^^^^^^^^
    |
    = help: add #![feature(rustc_private)] to the crate attributes to enable

error: use of unstable library feature 'rustc_private' (see issue #27812)
   --> /Users/manishearth/.cargo/registry/src/github.com-1ecc6299db9ec823/env_logger-0.4.2/src/lib.rs:233:22
    |
233 |         where F: Fn(&LogRecord) -> String + Sync + Send
    |                      ^^^^^^^^^
    |
    = help: add #![feature(rustc_private)] to the crate attributes to enable

error: use of unstable library feature 'rustc_private' (see issue #27812)
   --> /Users/manishearth/.cargo/registry/src/github.com-1ecc6299db9ec823/env_logger-0.4.2/src/lib.rs:259:42
    |
259 |     pub fn init(&mut self) -> Result<(), SetLoggerError> {
    |                                          ^^^^^^^^^^^^^^
    |
    = help: add #![feature(rustc_private)] to the crate attributes to enable

error: use of unstable library feature 'rustc_private' (see issue #27812)
   --> /Users/manishearth/.cargo/registry/src/github.com-1ecc6299db9ec823/env_logger-0.4.2/src/lib.rs:260:9
    |
260 |         log::set_logger(|max_level| {
    |         ^^^^^^^^^^^^^^^
    |
    = help: add #![feature(rustc_private)] to the crate attributes to enable

error: use of unstable library feature 'rustc_private' (see issue #27812)
   --> /Users/manishearth/.cargo/registry/src/github.com-1ecc6299db9ec823/env_logger-0.4.2/src/lib.rs:273:24
    |
273 |                 level: LogLevelFilter::Error,
    |                        ^^^^^^^^^^^^^^^^^^^^^
    |
    = help: add #![feature(rustc_private)] to the crate attributes to enable

error: use of unstable library feature 'rustc_private' (see issue #27812)
   --> /Users/manishearth/.cargo/registry/src/github.com-1ecc6299db9ec823/env_logger-0.4.2/src/lib.rs:304:29
    |
304 |     pub fn filter(&self) -> LogLevelFilter {
    |                             ^^^^^^^^^^^^^^
    |
    = help: add #![feature(rustc_private)] to the crate attributes to enable

error: use of unstable library feature 'rustc_private' (see issue #27812)
   --> /Users/manishearth/.cargo/registry/src/github.com-1ecc6299db9ec823/env_logger-0.4.2/src/lib.rs:307:24
    |
307 |             .unwrap_or(LogLevelFilter::Off)
    |                        ^^^^^^^^^^^^^^^^^^^
    |
    = help: add #![feature(rustc_private)] to the crate attributes to enable

error: use of unstable library feature 'rustc_private' (see issue #27812)
   --> /Users/manishearth/.cargo/registry/src/github.com-1ecc6299db9ec823/env_logger-0.4.2/src/lib.rs:310:30
    |
310 |     fn enabled(&self, level: LogLevel, target: &str) -> bool {
    |                              ^^^^^^^^
    |
    = help: add #![feature(rustc_private)] to the crate attributes to enable

error: use of unstable library feature 'rustc_private' (see issue #27812)
   --> /Users/manishearth/.cargo/registry/src/github.com-1ecc6299db9ec823/env_logger-0.4.2/src/lib.rs:325:34
    |
325 |     fn enabled(&self, metadata: &LogMetadata) -> bool {
    |                                  ^^^^^^^^^^^
    |
    = help: add #![feature(rustc_private)] to the crate attributes to enable

error: use of unstable library feature 'rustc_private' (see issue #27812)
   --> /Users/manishearth/.cargo/registry/src/github.com-1ecc6299db9ec823/env_logger-0.4.2/src/lib.rs:329:28
    |
329 |     fn log(&self, record: &LogRecord) {
    |                            ^^^^^^^^^
    |
    = help: add #![feature(rustc_private)] to the crate attributes to enable

error: use of unstable library feature 'rustc_private' (see issue #27812)
   --> /Users/manishearth/.cargo/registry/src/github.com-1ecc6299db9ec823/env_logger-0.4.2/src/lib.rs:330:13
    |
330 |         if !Log::enabled(self, record.metadata()) {
    |             ^^^^^^^^^^^^
    |
    = help: add #![feature(rustc_private)] to the crate attributes to enable

error: use of unstable library feature 'rustc_private' (see issue #27812)
   --> /Users/manishearth/.cargo/registry/src/github.com-1ecc6299db9ec823/env_logger-0.4.2/src/lib.rs:208:45
    |
208 |                 format!("{}:{}: {}", record.level(),
    |                                             ^^^^^
    |
    = help: add #![feature(rustc_private)] to the crate attributes to enable

error: use of unstable library feature 'rustc_private' (see issue #27812)
   --> /Users/manishearth/.cargo/registry/src/github.com-1ecc6299db9ec823/env_logger-0.4.2/src/lib.rs:209:32
    |
209 |                         record.location().module_path(), record.args())
    |                                ^^^^^^^^
    |
    = help: add #![feature(rustc_private)] to the crate attributes to enable

error: use of unstable library feature 'rustc_private' (see issue #27812)
   --> /Users/manishearth/.cargo/registry/src/github.com-1ecc6299db9ec823/env_logger-0.4.2/src/lib.rs:209:43
    |
209 |                         record.location().module_path(), record.args())
    |                                           ^^^^^^^^^^^
    |
    = help: add #![feature(rustc_private)] to the crate attributes to enable

error: use of unstable library feature 'rustc_private' (see issue #27812)
   --> /Users/manishearth/.cargo/registry/src/github.com-1ecc6299db9ec823/env_logger-0.4.2/src/lib.rs:209:65
    |
209 |                         record.location().module_path(), record.args())
    |                                                                 ^^^^
    |
    = help: add #![feature(rustc_private)] to the crate attributes to enable

error: use of unstable library feature 'rustc_private' (see issue #27812)
   --> /Users/manishearth/.cargo/registry/src/github.com-1ecc6299db9ec823/env_logger-0.4.2/src/lib.rs:262:23
    |
262 |             max_level.set(logger.filter());
    |                       ^^^
    |
    = help: add #![feature(rustc_private)] to the crate attributes to enable

error: use of unstable library feature 'rustc_private' (see issue #27812)
   --> /Users/manishearth/.cargo/registry/src/github.com-1ecc6299db9ec823/env_logger-0.4.2/src/lib.rs:326:31
    |
326 |         self.enabled(metadata.level(), metadata.target())
    |                               ^^^^^
    |
    = help: add #![feature(rustc_private)] to the crate attributes to enable

error: use of unstable library feature 'rustc_private' (see issue #27812)
   --> /Users/manishearth/.cargo/registry/src/github.com-1ecc6299db9ec823/env_logger-0.4.2/src/lib.rs:326:49
    |
326 |         self.enabled(metadata.level(), metadata.target())
    |                                                 ^^^^^^
    |
    = help: add #![feature(rustc_private)] to the crate attributes to enable

error: use of unstable library feature 'rustc_private' (see issue #27812)
   --> /Users/manishearth/.cargo/registry/src/github.com-1ecc6299db9ec823/env_logger-0.4.2/src/lib.rs:330:39
    |
330 |         if !Log::enabled(self, record.metadata()) {
    |                                       ^^^^^^^^
    |
    = help: add #![feature(rustc_private)] to the crate attributes to enable

error: use of unstable library feature 'rustc_private' (see issue #27812)
   --> /Users/manishearth/.cargo/registry/src/github.com-1ecc6299db9ec823/env_logger-0.4.2/src/lib.rs:335:42
    |
335 |             if !filter.is_match(&*record.args().to_string()) {
    |                                          ^^^^
    |
    = help: add #![feature(rustc_private)] to the crate attributes to enable

error: use of unstable library feature 'rustc_private' (see issue #27812)
   --> /Users/manishearth/.cargo/registry/src/github.com-1ecc6299db9ec823/env_logger-0.4.2/src/lib.rs:386:32
    |
386 |                     Err(_) => (LogLevelFilter::max(), Some(part0)),
    |                                ^^^^^^^^^^^^^^^^^^^
    |
    = help: add #![feature(rustc_private)] to the crate attributes to enable

error: use of unstable library feature 'rustc_private' (see issue #27812)
   --> /Users/manishearth/.cargo/registry/src/github.com-1ecc6299db9ec823/env_logger-0.4.2/src/lib.rs:389:47
    |
389 |             (Some(part0), Some(""), None) => (LogLevelFilter::max(), Some(part0)),
    |                                               ^^^^^^^^^^^^^^^^^^^
    |
    = help: add #![feature(rustc_private)] to the crate attributes to enable

error: Compilation failed, aborting rustdoc

error: Could not document `env_logger`.
Build failed, waiting for other jobs to finish...
error: build failed


command did not execute successfully: "/Users/manishearth/mozilla/rust/build/x86_64-apple-darwin/stage0/bin/cargo" "doc" "-j" "8" "--target" "x86_64-apple-darwin" "--release" "--manifest-path" "/Users/manishearth/mozilla/rust/src/rustc/Cargo.toml" "--features" " jemalloc" "-p" "rustc_driver" "-p" "rustdoc"
expected success, got: exit code: 101

```

</details>

**Comment by Manishearth** [comment]
cc @alexcrichton

**Comment by ollie27** [comment]
Fixed by #40852.
